### PR TITLE
add sessions for skipping time-costly visa validation on requests

### PR DESCRIPTION
**Pull request by teemukataja**
### Changes
- added `ristretto` cache for storing session keys and dataset permission lists
- added session key authentication based on `sda_session_key` cookie
  - successful authentication with `Bearer` token will set a session cookie to the response headers
  - a client can send these cookies back in request headers to skip `Bearer` token authentication
  - new config section for managing session cookies
- added `makecert.sh` helper script to generate self-signed TLS certificate for testing HTTPS and secure cookies

### Rationale
- it takes a long time to get and validate visas, so it's not feasible to do it for every request when reading files
- performance improvement on `localhost` with 20 visas was 1000x (from ~2s to 2ms)

### CODE REVIEW

File: internal/config/config.go [hunk 1]
```diff
@@ -53,6 +55,17 @@ type AppConfig struct {
 	ArchivePath string
 }
 
+type SessionConfig struct {
+	// Session key expiration time.
```
Comment (blankdots) [inline internal/config/config.go:59]: i think we should disable it by default, if the `Session_expiration` and `Session_domain` are not set

e.g. lines 126-128 or 218-220
```
if viper.IsSet("configFile") {
		viper.SetConfigFile(viper.GetString("configFile"))
}
```

Comment (teemukataja) [reply internal/config/config.go]: sessions disabled by default, and also fixed a time-to-int conversion bug that this made apparent (`time.Duration` is in nanoseconds, but cookie age is in seconds, so cookie lifetime got accidentally increased by a factor of `1e9`)

File: test/makecert.sh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+# # if you need a new self-signed cert to test https
+openssl genrsa -out "tls.key" 2048
```
Comment (blankdots) [inline test/makecert.sh:4]:
```suggestion
openssl genrsa -out "tls.key" 4096
```

File: test/makecert.sh [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+openssl req -new -key "tls.key" -out "tls.csr" -subj '/C=FI/ST=Uusimaa/L=Espoo/O=CSC/CN=localhost'
+openssl x509 -req -days 365 -in "tls.csr" -signkey "tls.key" -out "tls.crt"
```
Comment (blankdots) [inline test/makecert.sh:6]:
```suggestion
openssl x509 -req -days 7300 -in "tls.csr" -signkey "tls.key" -out "tls.crt"
```
to keep consistent with: https://github.com/neicnordic/sda-pipeline/blob/master/dev_utils/make_certs.sh

File: internal/session/session.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	if err != nil {
+		log.Errorf("failed to create session cache, reason=%v", err)
+		return nil, err
+	}
+	log.Debug("session cache created")
+	return sessionCache, nil
+}
+
+// Get returns a value from cache at key
+// DON'T USE DEBUG LOGS IN PRODUCTION, AS SESSION KEY WILL BE LOGGED
```
Comment (blankdots) [inline internal/session/session.go:44]: I don't think we need to log it even for debug. We might get an alert as: https://github.com/neicnordic/sda-pipeline/security/code-scanning/3?query=ref%3Arefs%2Fheads%2Fmaster so we can just remove that and add it when we develop

Comment (teemukataja) [reply internal/session/session.go]: removed

**Review by blankdots — APPROVED** [review]

**Pull request merged**